File: leetcode/0984-string-without-aaa-or-bbb/solution.cpp
#include <string>
using namespace std;

class Solution {
public:
    string strWithout3a3b(int a, int b) {
        string result;

        while (a > 0 || b > 0) {
            int n = (int)result.size();
            bool lastTwoSame = n >= 2 && result[n - 1] == result[n - 2];

            if (lastTwoSame) {
                if (result[n - 1] == 'a') {
                    result += 'b';
                    b--;
                } else {
                    result += 'a';
                    a--;
                }
            } else {
                if (a >= b && a > 0) {
                    result += 'a';
                    a--;
                } else {
                    result += 'b';
                    b--;
                }
            }
        }

        return result;
    }
};

File: leetcode/0984-string-without-aaa-or-bbb/Solution.java
class Solution {
    public String strWithout3a3b(int a, int b) {
        StringBuilder result = new StringBuilder();

        while (a > 0 || b > 0) {
            int n = result.length();
            boolean lastTwoSame = n >= 2 && result.charAt(n - 1) == result.charAt(n - 2);

            if (lastTwoSame) {
                if (result.charAt(n - 1) == 'a') {
                    result.append('b');
                    b--;
                } else {
                    result.append('a');
                    a--;
                }
            } else {
                if (a >= b && a > 0) {
                    result.append('a');
                    a--;
                } else {
                    result.append('b');
                    b--;
                }
            }
        }

        return result.toString();
    }
}
